src/thunderbots/software/ai/passing/pass_generator.h

```diff
@@ -0,0 +1,249 @@
+#include "ai/world/world.h"
+#include "util/gradient_descent.h"
+#include "util/parameter/dynamic_parameters.h"
+#include "util/timestamp.h"
+
+namespace AI::Passing
+{
+    /**
+     * This class is responsible for generating passes for us to perform
+     *
+     * == General Description ==
+     * It is built such that when it is constructed, a thread is immediately started in
+     * the background continuously optimizing/pruning/re-generating passes. As such, any
+     * modifications to data in this class through a public interface *must* be managed
+     * by mutexes. Generally in functions that touch data we use "std::lock_guard" to
+     * take ownership of the mutex protecting that data for the duration of the function.
+     *
+     * == Making Changes/Additions ==
+     * Whenever you change/add a function, you need to ask: "what data is this _directly_
+     * touching?". If the function is touching anything internal to the class, you should
+     * be getting a lock on the mutex for that data member for the duration of the
+     * function (see below for examples). Note the *directly* bit! If you are
+     * changing/adding function "A", and you have it call function "B", if B touches
+     * data, then B should be responsible for getting a lock on the data. If you acquire
+     * a lock in A, then call B, which also requires a lock, then the threads will
+     * deadlock and everything will grind to a halt.
+     *
+     */
+    class PassGenerator
+    {
+       public:
+        // Delete the default constructor, we want to force users to choose what
+        // pass quality they deem reasonable
+        PassGenerator() = delete;
+
+        // Delete the copy and assignment operators because this class really shouldn't
+        // need them and we don't want to risk doing anything nasty with the internal
+        // threading this class uses
+        PassGenerator& operator=(const PassGenerator&) = delete;
+        PassGenerator(const PassGenerator&)            = delete;
+
+        /**
+         * Create a PassGenerator with given parameters
+         *
+         * @param min_reasonable_pass_quality A value in [0,1] representing the minimum
+         *                                    quality for a pass to be considered
+         *                                    "reasonable", with higher being better
+         */
+        explicit PassGenerator(double min_reasonable_pass_quality);
+
+        /**
+         * Updates the world
+         *
+         * @param world
+         */
+        void setWorld(World world);
+
+        /**
+         * Updates the point that we are passing from
+         *
+         * @param passer_point the point we are passing from
+         */
+        void setPasserPoint(Point passer_point);
+
+        /**
+         * Set the target region that we would like to pass to
+         *
+         * @param area The area to pass to
+         */
+        void setTargetRegion(Rectangle area);
```

Comment:
Should we also have a `clearTargetRegion` function? Right now once we set the target region it will never be `std::nullopt` again and always incorporated in the calculation. The only way to get back to evaluating over the entire field would be to pass the entire field area as a rectangle.

Reply:
Changed this take an `std::optional<Rectangle>`, so the user can specify no region by giving it `std::nullopt`. Changed the javadoc comment to reflect this.